### DISCUSSION TITLE
Pyright cannot detect condition context from property

### DISCUSSION THREAD

**Discussion by catcombo** [microsoft/pyright#9100] [Q&A]
Hi folks!

Don't know if it's a bug or not. Here is the code example:
```python
class Content:
    media_type: str


class ResponseView:
    def __init__(self, content: Content | None):
        self.content = content

    @property
    def has_content(self) -> bool:
        return self.content is not None

    @property
    def media_type(self) -> str | None:
        return self.content.media_type if self.has_content else None
```
pyright complains
![image](https://github.com/user-attachments/assets/8e77d919-556d-4825-b730-e22a44a25a3b)
But if I inline condition from `has_content` property then it stops complaining
![image](https://github.com/user-attachments/assets/4ff21873-b176-4d9a-a9ed-0d620e461d33)
Is it a bug? WDYT?

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

The `has_content` property is checking whether `self.content` is `None`, but this fact isn't visible outside of the `has_content` body. Static analysis tools like pyright, mypy, etc. do not propagate knowledge like this across function calls. 

If you want your code to work with static analysis, you'll need to avoid this pattern. Here's one way that you could change your code to eliminate type errors:

```python
    @property
    def media_type(self) -> str | None:
        # Check self.content directly rather than calling a method to check it
        return self.content.media_type if self.content else None
```

**Reply by catcombo** [reply]
Thanks for the detailed explanation, Eric!
